libsolutil/CDCL.h

```diff
@@ -48,16 +49,137 @@ struct Literal
 		return std::make_tuple(positive, variable) < std::make_tuple(_other.positive, _other.variable);
 	}
 };
-using Clause = std::vector<Literal>;
+
+struct Clause {
+	const Literal& front() const
+	{
+		return lits.front();
+	}
+	const Literal& back() const
+	{
+		return lits.back();
+	}
+	Literal& front()
+	{
+		return lits.front();
+	}
+	Literal& back()
+	{
+		return lits.back();
+	}
+	const Literal& operator[](const size_t at) const
+	{
+		return lits.at(at);
+	}
+	const Literal& at(const size_t at) const
+	{
+		return lits.at(at);
+	}
+	Literal& operator[](const size_t at)
+	{
+		return lits.at(at);
+	}
+	Literal& at(const size_t at)
+	{
+		return lits.at(at);
+	}
+	auto size() const
+	{
+		return lits.size();
+	}
+	auto empty() const
+	{
+		return lits.empty();
+	}
+	auto begin() {
+		return lits.begin();
+	}
+	auto end() {
+		return lits.end();
+	}
+	auto begin() const {
+		return lits.begin();
+	}
+	auto end() const {
+		return lits.end();
+	}
+	void push_back(const Literal& l)
+	{
+		lits.push_back(l);
+	}
+
+	std::vector<Literal> lits;
+	uint64_t ID;
+};
+
+class TriState {
+public:
+	TriState(const bool b)
+	{
+		if (b)
+			val = 1;
+		else
+			val = 0;
+	}
+
+	TriState()
+	{}
+
+	bool operator==(const TriState& other) const
+	{
+		return val == other.val;
+	}
+
+	bool operator!=(const TriState& other) const
+	{
+		return val != other.val;
+	}
+
+	bool toBool() const {
+		if (val == 1)
+			return true;
+		else if (val == 0)
+			return false;
+		else
+		{
+			assert(val== 2);
+			assert(false && "UNSET cannot be converted to bool");
+		}
+	}
+
+	std::string toString()
+	{
+		if (val == 1)
+			return "true";
+		else if (val == 0)
+			return "false";
+		else
+		{
+			assert(val == 2);
+			return "unset";
+		}
+	}
+
+
+private:
+	// Default value is UNSET
+	char val = 2;
+};
+
+const TriState trisate_true = TriState(true);
+const TriState tristate_false = TriState(false);
```

Comment:
Not needed

Reply:
That's true. I am using it the code as a helper, that what we are checking is a tristate value. For example:

```
if (value(clause[1]) == TriState::t_true())
{
  // Clause is already satisfied, keep the watch.		}
```

Would otherwise be:

```
if (value(clause[1]) == true)
{
  // Clause is already satisfied, keep the watch.		}
```

Which is weird, because one would think that it's possible to then do:

```
if (value(clause[1]))
{
  // Clause is already satisfied, keezp the watch.		}
```

but for that, we'd need a a `bool` conversion to `TriState`:

```
class TriState{
[...]
	explicit operator bool() const
	{
		if (val == 0) return false;
		else if (val == 1) return true;
		else assert(false);
	}
}
```

But we need to have that `assert()` there, or we could face some serious issues. However, that assert will fire when it's `TriState::t_unset` which it will be, quite often. The reason why we need the `assert` is that if we don't have the assert, then we can only return `true` or `false`, and if we return `false` for both `TriState::t_false()` and `TriState::t_unset()`, then double negation will not be the same as no negation, which can be very unexpected for the programmer.

So overall, as far as I can see, it may be best to compare to the now `static constexpr` member function `TriState::t_true()`? What do you think?

Reply:
I agree, the conversion to bool would be dangerous and maybe `value(clause[1]) == true` could look weird, but I would say that in this case, we should also not have an implicit conversion from bool, but only an explicit one.

Reply:
Yeah, I think explicit conversion from `true`/`false` would be nicer, too! I just added it, it's a total of 3-line change (1f5445289c835cec534a17b85c2be59a20dea632), and I think it's well worth it. Let me know what you think!